### ISSUE TITLE
pml边界实现问题

### ISSUE THREAD

**Issue by LZT1945** [ar4/deepwave#94]
Hi Dr. Richardson,
目前我在学习您的正演程序，请问在运行您的scaler程序时，pml_freq和pml_width时如何工作并消除反射波的

**Comment by ar4** [comment]
Hi @LZT1945,

These are described in the [documentation of the propagator](https://ausargeo.com/deepwave/usage#id4). Deepwave uses a PML to create absorbing boundaries. `pml_freq` is the frequency that is passed to the method, which you will normally set to the dominant frequency of your source, and `pml_width` is the number of cells wide to create the PML, which affects how effective it is. As noted in the documentation, the scalar propagator uses [the PML method of Pasalic and McGarry](https://doi.org/10.1190/1.3513453).

**Comment by LZT1945** [comment]
这是我的代码，
import torch
import matplotlib.pyplot as plt
import numpy as np
import math
# 设备选择：使用CUDA（如果可用）
plt.rcParams['axes.unicode_minus'] = False # 显示负号
plt.rcParams['font.sans-serif'] = ['SimHei']
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

torch.set_default_dtype(torch.float32)

# 参数设置
nx, nz = 200,200
nt = 3000
dx = dz = 10.0
dt = 0.001
pml_width = 10            # PML宽度（加厚减少反射）
fd_order = 8              # 差分阶数（支持2、4、6、8）
use_cpml = True           # 是否使用CPML（若为False，则关闭PML）
freq = 3.0                # 主频
pml_freq = 1.5 * freq      # 用于 PML 的主频率（提升以匹配子波频谱）
save_nt =100
# 声速模型 (m/s)，二维张量，需为叶子节点以支持梯度
v_base = torch.ones(nz, nx, device=device) * 1500.0
v_base[100:, :] = 2500.0

v = v_base.clone().detach().requires_grad_()

# 源和接收器位置
src_x, src_z = 100, 11
rec_x_list = list(range(0, nx, 1))
rec_z = 11
# 可视化速度模型
plt.figure(figsize=(6, 6))
plt.imshow(v.detach().cpu().numpy(), cmap='jet', extent=[0, nx * dx, nz * dz, 0])
plt.colorbar(label='v (m/s)')
plt.title('v_model')
plt.xlabel('x (m)')
plt.ylabel('z (m)')
plt.tight_layout()
plt.show()

# 构造 Ricker 子波
f0 = freq
t = torch.arange(0, nt*dt, dt, device=device)
peak_time = 1.5 / f0
tau = t - peak_time
ricker = (1 - 2 * (math.pi * f0 * tau) ** 2) * torch.exp(-(math.pi * f0 * tau) ** 2)

# 初始波场（开启 requires_grad 以便反传）
u = torch.zeros(nz, nx, device=device)
u_past = torch.zeros_like(u)
u_future = torch.zeros_like(u)

# 存储接收器记录
seismogram = torch.zeros(len(rec_x_list), nt, device=device)


snapshots = []
if use_cpml:
    R = 1e-8
    order = 3
    delta = pml_width * dx
    freq = f0

    # Calculate the maximum damping based on frequency
    sigma_max = 3.0 * 2 * 3.1415926 * freq * torch.log(torch.tensor(1.0 / R))
    a_max = 0.1
    k_max = 1 + 5 * (0.5 / f0)

    sigma_x = torch.zeros(nx, device=device)
    sigma_z = torch.zeros(nz, device=device)
    kappa_x = torch.ones(nx, device=device)
    kappa_z = torch.ones(nz, device=device)
    alpha_x = torch.zeros(nx, device=device)
    alpha_z = torch.zeros(nz, device=device)

    for i in range(pml_width):
        x_val = (pml_width - i) / pml_width
        smooth_val = 1 - math.exp(-order * x_val)

        sigma_x[i] = sigma_x[-i - 1] = sigma_max * smooth_val
        sigma_z[i] = sigma_z[-i - 1] = sigma_max * smooth_val
        smooth_alpha = (1 - x_val) ** order  # 使用更平滑的衰减函数

        alpha_x[i] = alpha_x[-i - 1] = a_max * smooth_alpha
        alpha_z[i] = alpha_z[-i - 1] = a_max * smooth_alpha
        kappa_x[i] = kappa_x[-i - 1] = 1 + (k_max - 1) * smooth_val
        kappa_z[i] = kappa_z[-i - 1] = 1 + (k_max - 1) * smooth_val

    Sigma_x, Sigma_z = torch.meshgrid(sigma_z, sigma_x, indexing='ij')
    Kappa_x, Kappa_z = torch.meshgrid(kappa_z, kappa_x, indexing='ij')
    Alpha_x, Alpha_z = torch.meshgrid(alpha_z, alpha_x, indexing='ij')
    phi_x = torch.zeros(nz, nx, device=device)
    phi_z = torch.zeros(nz, nx, device=device)

    # Visualization of PML Damping Parameters
    import matplotlib.pyplot as plt
    plt.figure(figsize=(12, 4))
    plt.subplot(1, 3, 1)
    plt.plot(sigma_x.cpu()); plt.title("σ_x")
    plt.subplot(1, 3, 2)
    plt.plot(alpha_x.cpu()); plt.title("α_x")
    plt.subplot(1, 3, 3)
    plt.plot(kappa_x.cpu()); plt.title("κ_x")
    plt.tight_layout()
    plt.show()
# 高阶差分拉普拉斯

def laplacian_fd(u):
    coeffs = {
        2: [1, -2, 1],
        4: [-1/12, 4/3, -2.5, 4/3, -1/12],
        6: [1/90, -3/20, 3/2, -49/18, 3/2, -3/20, 1/90],
        8: [-1/560, 8/315, -1/5, 8/5, -205/72, 8/5, -1/5, 8/315, -1/560]
    }
    c = coeffs[fd_order]
    lap = torch.zeros_like(u)
    half = len(c) // 2
    for i in range(-half, half+1):
        lap += c[i+half] * (torch.roll(u, shifts=i, dims=0) + torch.roll(u, shifts=i, dims=1))
    return lap / (dx * dx)

# 时间推进循环
for it in range(nt):
    u = u.clone()
    if it <= len(ricker):
        u[src_z, src_x] = u[src_z, src_x] + ricker[it]
    lap = laplacian_fd(u)

    if use_cpml:
        phi_x = (phi_x - dt * Alpha_x * phi_x) + dt / Kappa_x * Sigma_x * lap
        phi_z = (phi_z - dt * Alpha_z * phi_z) + dt / Kappa_z * Sigma_z * lap
        # damping = torch.exp(-(Sigma_x + Sigma_z + Alpha_x + Alpha_z) * dt / Kappa_x)
        damping_x = torch.exp(-dt * (Sigma_x / Kappa_x + Alpha_x))
        damping_z = torch.exp(-dt * (Sigma_z / Kappa_z + Alpha_z))
        damping = damping_x * damping_z
        u_future = damping * (2 * u - u_past + (v * dt) ** 2 * (lap + phi_x + phi_z))

    else:
        u_future = 2 * u - u_past + (v * dt) ** 2 * lap
        u_future[0, :] = 0
        u_future[-1, :] = 0
        u_future[:, 0] = 0
        u_future[:, -1] = 0

    for i, rx in enumerate(rec_x_list):
        seismogram[i, it] = u[rec_z, rx]

    if it % save_nt == 0:
        snapshots.append(u.detach().clone().cpu())

    u_past, u = u, u_future

loss = seismogram.pow(2).sum()
loss.backward()
grad_v = v.grad.cpu()

# 可视化地震记录
plt.rcParams['font.sans-serif'] = ['SimHei']
plt.figure(figsize=(10, 5))
plt.imshow(seismogram.cpu().detach().numpy().T, cmap='gray', aspect='auto',
           extent=[0, len(rec_x_list)*dx,nt*dt, 0])
plt.xlabel('channel (m)')
plt.ylabel('time (s)')
plt.title('Composite seismogram')
plt.colorbar(label='amplitude')
plt.show()

# 可视化梯度
plt.figure(figsize=(6, 6))
grad_v_np = grad_v.detach().numpy()
grad_v_norm = grad_v_np / (np.max(np.abs(grad_v_np)) + 1e-10)
plt.imshow(grad_v_norm, cmap='seismic', extent=[0, nx*dx, nz*dz, 0], vmin=-1, vmax=1)
plt.title("gard")
plt.colorbar(label='归一化梯度')
plt.xlabel('x (m)')
plt.ylabel('z (m)')
plt.show()



# 快照统一色标 + 合并显示
snap_min = min(s.min() for s in snapshots)
snap_max = max(s.max() for s in snapshots)

nrow = int(np.ceil(len(snapshots) / 4))
fig, axes = plt.subplots(nrow, 4, figsize=(16, 3 * nrow))
axes = axes.flatten()
for i, snap in enumerate(snapshots):
    im = axes[i].imshow(snap.numpy(), cmap='seismic', extent=[0, nx*dx, nz*dz, 0], vmin=snap_min, vmax=snap_max)
    axes[i].set_title(f't={i*save_nt*dt:.3f}s')
    axes[i].set_xlabel('x (m)')
    axes[i].set_ylabel('z (m)')
for j in range(len(snapshots), len(axes)):
    axes[j].axis('off')
fig.suptitle("波场快照", fontsize=16)
fig.colorbar(im, ax=axes.tolist(), shrink=0.6, label='振幅')
plt.tight_layout()
plt.show()

# 可视化第 i 个接收器
rec_index = 100  # 可修改为其他接收器索引
_, ax = plt.subplots(1, 2, figsize=(18, 18))
trace = seismogram[rec_index].cpu().detach().numpy()
ax[0].plot(trace, t.cpu().numpy())
ax[0].set_ylim(ymin = 0, ymax = 1)
ax[0].invert_yaxis()
ax[0].set_title("Receiver data")
ax[0].set_xlabel("time")  # 时间轴
ax[0].set_ylabel("amplitude")
ax[0].set_xlim(xmin = -40, xmax = 40)
在使用pml边界时，效果并不理想

![Image](https://github.com/user-attachments/assets/f4c0640f-6408-4664-935f-2dea33fd3a4d)

**Comment by ar4** [comment]
Hello,

You appear to be asking for help with your own wave propagator, rather than
for assistance using Deepwave. I unfortunately do not have time to study
your code with the care needed to identify any possible mistakes, but I
suggest that if your PML is not adequately absorbing incident waves, then
it might be because it is not thick enough. It looks like you are using a
wave with a peak frequency of 3 Hz in a medium with a wave speed of 2500
m/s. The peak wavelength will therefore be 833 m. With a grid spacing of 10
m and a PML only 10 cells wide, your PML will only be 100 m wide. You may
therefore like to try increasing the width of the PML so that it is a
larger fraction of the wavelength.

**Comment by LZT1945** [comment]
非常感谢您的回复

**Comment by ar4** [comment]
I also see some other possible problems. For example, your value of R
(1e-8) is unusual (1e-3 to 1e-5 is more typical), and you have used a
hardcoded value of 0.1 for a_max instead of what I think is the more
typical pi times the dominant frequency. It is not easy to get the PML
right, but I hope you are able to resolve the problem you encountered.

**Comment by ar4** [comment]
I'm going to close this issue now, but please feel free to reopen it or to create a new one if you have any other questions.
